### DISCUSSION TITLE
Using fp-ts Array's isNonEmpty with fp-ts-std Predicate's when

### DISCUSSION THREAD

**Discussion by ammarrik** [gcanti/fp-ts#1821] [Q&A]
Can somebody please help me to use fp-ts-std when with fp-ts Array's isNonEmpty?

I'm currently using it like this;
```js
import { when } from "fp-ts-std/Predicate"
import { isNonEmpty } from "fp-ts/Array"

when(isNonEmpty([]), do something)
```

but it is displaying an error;
`Cannot assign boolean to predicate`

**Comment by samhh** [comment]
I think you mean the [`when` in fp-ts-std/Function](https://samhh.github.io/fp-ts-std/modules/Function.ts.html#when).

`isNonEmpty([])` isn't a predicate, it's a boolean. Remove the `([])` application and that part will typecheck. You'll also need to shift the `do something` as everything in fp-ts-std is curried.

Example:

```ts
import * as A from 'fp-ts/Array'
import { when } from 'fp-ts-std/Function'

const f: Endomorphism<string> = when(A.isNonEmpty)(A.append('another item for my non empty array'))
```

NB as currently written `when` won't carry through the refinement of `isNonEmpty`.

**Reply by ammarrik** [reply]
this is what I am doing but this also throws error:
```js
const mapping = map((body) => ({ value: body.id, label: body.classname }))
const classOption = when(isNonEmpty)(mapping)(classesQuery.data ?? [])
```

`Argument of type 'IClass[]' is not assignable to parameter of type '{ value: any; label: any; }[]'.`

**Reply by samhh** [reply]
Add an explicit type annotation to `mapping`.

**Reply by ammarrik** [reply]
```js
const mapping: (body: IClass[]) => ISelect[] = map((body) => ({ value: body.id, label: body.classname }))
const classOption = when(isNonEmpty)(mapping)(classesQuery.data ?? [])
```
now `mapping` assigned to `when`, says;
`Argument of type '(body: IClass[]) => ISelect[]' is not assignable to parameter of type 'Endomorphism<IClass[]>'`

btw thank you for your help

**Reply by samhh** [reply]
np 🙂 

`when` needs the mapping function to be an endomorphism i.e. same input and output type. That way the result is the same regardless of whether or not the condition matches (whether or not the function is applied).
